File: text_cloud.py
import sys
import re

#Jonathan Burdette
#Text Cloud Implementation

if len(sys.argv) < 4:
    print("Invalid syntax. To run program, enter: \"python text_cloud.py <input file name> <exclude file name> <output file name>\"")
else:
    input_file = sys.argv[1]
    exclude_file = sys.argv[2]
    output_file = sys.argv[3]

    #store words from exclude file in dict (blank values)
    exclude_dict = {}
    f = open(exclude_file, "r")
    lines = f.readlines()
    for line in lines:
        words = re.split("[^A-Z'a-z]+", line)
        for word in words:
            exclude_dict[word] = ""

    #store the words and their counts from input file
    input_dict = {}
    f = open(input_file, "r", encoding="utf8")
    lines = f.readlines()
    for line in lines:
        words = re.split("[^A-Z'a-z]+", line)
        for word in words:
            word = word.lower()
            if exclude_dict.get(word) == None and re.match("[a-zA-Z]", word) and len(word) > 1:
                count = input_dict.get(word)
                if count == None:
                    input_dict[word] = 1
                else:
                    input_dict[word] = count + 1
               
    #sort the words based on count
    sorted_counts = sorted(input_dict.items(), reverse=True, key=lambda x: x[1])

    #take the top fifty words based on count
    top_fifty = sorted_counts[0:50]

    #sort the top fifty in alphabetical order
    top_fifty.sort()

    #info for HTML file
    range = (sorted_counts[0][1] - sorted_counts[49][1])
    size_factor = 1000 / range
    colors = ["#623224", "#FABE8B", "#EC1D23", "#0160B0"]

    f = open(output_file, "w") 
    color_count = 0
    for var in top_fifty:
        if color_count == 4: #colors restart after 4 iterations
            color_count = 0

        fontsize = var[1] * size_factor
        f.write("<span style=\"font-size:")
        f.write(str(fontsize))
        f.write("%; color:")
        f.write(colors[color_count])
        f.write(";\">")
        f.write(var[0])
        f.write("</span> &nbsp; &nbsp;\n")
        color_count = color_count + 1

File: textCloud.js
const fs = require('fs');

//Jonathan Burdette
//Text Cloud Implementation
//I used an answer from this on line 45 https://stackoverflow.com/questions/37982476/how-to-sort-a-map-by-value-in-javascript

if(process.argv.length < 5) {
    console.log("Invalid syntax. To run program, enter: \"node textCloud.js <input file name> <exclude file name> <output file name>\"");
} else {
    var inputFile = process.argv[2];
    var excludeFile = process.argv[3];
    var outputFile = process.argv[4];

    //store words from exclude file in map (blank values)
    var excludeMap = new Map();
    fs.readFile(excludeFile, function(err, lines) {
        if(err) throw err;

        var wordArr = lines.toString().split(/[^A-Z'a-z]/);
        for(var i in wordArr) {
            var word = wordArr[i].toLowerCase();
            excludeMap.set(word, "");
        }
    })

    //store the words and their counts from input file
    var inputMap = new Map();
    fs.readFile(inputFile, function(err, lines) {
        if(err) throw err;

        var wordArr = lines.toString().split(/[^A-Z'a-z]/);
        for(var i in wordArr) {
            var word = wordArr[i].toLowerCase();
            if(!excludeMap.has(word) && word.match(/[A-Za-z]/) && word.length > 1) {
                var count = inputMap.get(word);
                if(count == null) {
                    inputMap.set(word, 1);
                } else {
                    inputMap.set(word, count + 1);
                }
            }
        }

        //sort the words based on count
        var sortedCounts = [...inputMap.entries()].sort(function(a, b){return b[1] - a[1]});

        //take the top fifty words based on count
        var topFifty = sortedCounts.slice(0, 50);

        //calculate info for html file based on top fifty words
        var range = (topFifty[0][1] - topFifty[49][1]);
        var sizeFactor = (1000 / range);

        //sort the top fifty in alphabetical order
        var sortedWords = topFifty.sort();

        //loop through the sorted words to create a string with html info
        var colors = ["#623224", "#FABE8B", "#EC1D23", "#0160B0"];
        var count = 0;
        var htmlLines = "";
        for(var i in sortedWords) {
            if(count == 4) { //colors restart after 4 iterations
                count = 0;
            }
            var fontSize = sortedWords[i][1] * sizeFactor;
            htmlLines = htmlLines + "<span style=\"font-size:"+fontSize+"%; color:"+colors[count]+";\">"+sortedWords[i][0]+"</span> &nbsp; &nbsp;\n";
            count++;
        }

        //write html file
        fs.writeFile(outputFile, htmlLines, function(err) {
            if(err) throw err;
        })
    })
}
